src/utils/transform.ts

```diff
@@ -1,20 +1,94 @@
 import * as ESTree from 'estree';
-import { Options as AcornOptions } from 'acorn';
 import { decode } from 'sourcemap-codec';
 import { locate } from 'locate-character';
 import error from './error';
 import getCodeFrame from './getCodeFrame';
 import Graph from '../Graph';
-import { defaultAcornOptions } from '../Module';
 import {
 	Plugin,
 	RollupWarning,
 	RollupError,
 	SourceDescription,
+	PluginContext,
 	RawSourceMap
 } from '../rollup/types';
 import Program from '../ast/nodes/Program';
-import { TransformContext } from '../rollup/types';
+
+function augmentCodeLocation<T extends RollupError | RollupWarning>({
+	object,
+	pos,
+	code,
+	id,
+	source,
+	pluginName
+}: {
+	object: T;
+	pos: { line: number; column: number };
+	code: string;
+	id: string;
+	source: string;
+	pluginName: string;
+}): T {
+	if (object.code) object.pluginCode = object.code;
+	object.code = code;
+
+	if (pos !== undefined) {
+		if (pos.line !== undefined && pos.column !== undefined) {
+			const { line, column } = pos;
+			object.loc = { file: id, line, column };
+			object.frame = getCodeFrame(source, line, column);
+		} else {
+			object.pos = <any>pos;
+			const { line, column } = locate(source, pos, { offsetLine: 1 });
+			object.loc = { file: id, line, column };
+			object.frame = getCodeFrame(source, line, column);
+		}
+	}
+
+	object.plugin = pluginName;
+	object.id = id;
+
+	return object;
+}
+
+function createPluginTransformContext(
+	graph: Graph,
+	plugin: Plugin,
+	id: string,
+	source: string
+): PluginContext {
+	return {
+		isExternal: graph.pluginContext.isExternal,
+		resolveId: graph.pluginContext.resolveId,
+		parse: graph.pluginContext.parse,
+		emitAsset: graph.pluginContext.emitAsset,
+		setAssetSource: graph.pluginContext.setAssetSource,
+		warn(warning: RollupWarning | string, pos?: { line: number; column: number }) {
+			if (typeof warning === 'string') warning = { message: warning };
+			warning = augmentCodeLocation({
+				object: warning,
+				pos,
+				code: 'PLUGIN_WARNING',
+				id,
+				source,
+				pluginName: plugin.name
```

Comment:
If a plugin does not have a name (which happens often for ad-hoc plugins) the warning will read: "(!) undefined plugin:  ..." which is somewhat confusing. One approach could be to set the `pluginName` to "unnamed" which makes for less confusing warnings.

The same would apply to errors though in that case, the error message is less confusing (the message seems to assume this is not a plugin error). Still, I think if the error reads "[!] (unnamed plugin) Error: ...", this would provide the helpful information that the error did not originate from rollup itself but probably from some code the user wrote themselves.

Reply:
Another issue I stumbled upon: When I generate warnings or errors from the `transformBundle` hook, I get a maximum callstack error instead.